# Cleaning The Data

Made by: Alexander Beaucage

Date: June 23 2023

Contact Info: Beaucagealex202@gmail.com

The goal of this notebook is to clean up the data set and get it ready for making a recommendation system.

In [1]:
# Imports
import numpy as np
import pandas as pd

In [2]:
# Getting the directory of the data to read it into pandas
datadir = r"C:\Users\Alexander\Documents\Data Science Boot Camp\Capstone\Copys\spotify_dataset.csv"

# Reading in the data
data = pd.read_csv(datadir)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_18232\583834019.py:5: DtypeWarning: Columns (4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv(datadir)


In [3]:
# What is with the extra columns????
data.head()
# I was expecting only four columns but got an extra 120 columns.

user_id                      "artistname"  \
0  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   
1  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
2  9cc0cfd4d7d7885102480dd99e7a90d6                      Tiffany Page   
3  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
4  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   

                                         "trackname"  "playlistname"  \
0               (The Angels Wanna Wear My) Red Shoes  HARD ROCK 2010   
1  (What's So Funny 'Bout) Peace, Love And Unders...  HARD ROCK 2010   
2                                   7 Years Too Late  HARD ROCK 2010   
3                              Accidents Will Happen  HARD ROCK 2010   
4                                             Alison  HARD ROCK 2010   

  Unnamed: 4 Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8 Unnamed: 9  ...  \
0        NaN        NaN        NaN        NaN        NaN        NaN  ...   
1        NaN        NaN        NaN        NaN        NaN        NaN  ...   
2        NaN        NaN        NaN        NaN        NaN        NaN  ...   
3        NaN        NaN        NaN        NaN        NaN        NaN  ...   
4        NaN        NaN        NaN        NaN        NaN        NaN  ...   

  Unnamed: 114 Unnamed: 115 Unnamed: 116 Unnamed: 117 Unnamed: 118  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

  Unnamed: 119 Unnamed: 120 Unnamed: 121 Unnamed: 122 Unnamed: 123  
0          NaN          NaN          NaN          NaN          NaN  
1          NaN          NaN          NaN          NaN          NaN  
2          NaN          NaN          NaN          NaN          NaN  
3          NaN          NaN          NaN          NaN          NaN  
4          NaN          NaN          NaN          NaN          NaN  

[5 rows x 124 columns]

In [4]:
# How many rows and columns?
data.shape
# 1048575 rows
# 124 columns(Should only be 4)

(1048575, 124)

Some weird stuff going on with the data.

First off is the extra columns.

Then there is the quotes around `artistname`, `trackname`, and `playlistname`.

Here is the plan of action for removing the extra columns
- Find and drop the rows with anything in the extra 120 columns
 1. Create a boolean mask for the rows with anything in the extra columns.
 2. Drop the things in that maks or even make a new data frame.
 
 
- Then drop the extra 120 columns
 1. drop the extra columns

In [5]:
# Create a boolean mask for if there is any entry in the extra columns
# The .all makes sure to combine the booleans into a single row insted of a array of booleans
boolmask = data.iloc[:,4:,].isna().all(axis = 1)

This seems backwards, but if the value is `True` that means that there is nothing in the extra columns for that row.

In [6]:
# Make sure the boolean mask looks right
boolmask

0          True
1          True
2          True
3          True
4          True
           ... 
1048570    True
1048571    True
1048572    True
1048573    True
1048574    True
Length: 1048575, dtype: bool

In [7]:
# Remove the rows with data in the extra columns
data.drop(boolmask[boolmask == False].index, inplace= True)

In [8]:
# Make data equal to the first 4 columns of data.
# Effectivly removing the extra columns
data = data.iloc[:,0:4]

In [9]:
# Take a look at the data to make sure it look right.
data.sample(5)

user_id  "artistname"       "trackname"  \
260016   552e2c862b95207bd5ee72f9de9820c6  Public Enemy   Fight The Power   
861853   d7d48e1d55970b769c04ce63ee2005b5  Soundflowers     Secret Wonder   
488785   3dad1058688d39a8ae96710e2501b26a      Jamie xx       Loud Places   
1042500  ab4d3778e65e1342be1ac754db166c8a         Finch       Post Script   
533446   1435064e3d682b77235ef67451be8474        KONGOS  Come With Me Now   

                         "playlistname"  
260016          THE ILLEST 90s MIX EVER  
861853   Intelligent Music Favorites Vo  
488785                     Office Beats  
1042500                          E.E.P.  
533446                      summer 2k14

Now it's time to remove those weird quotes.

In [10]:
# See what the column names actualy are.
data.columns

Index(['user_id', ' "artistname"', ' "trackname"', ' "playlistname"'], dtype='object')

In [11]:
# Change the column names to use be able to use them later.
data.rename(columns = {' "artistname"': 'artistname',
                       ' "trackname"': 'trackname', 
                       ' "playlistname"': 'playlistname'},
            inplace= True)

In [12]:
# Take a look to see if it worked
data.head()

user_id                        artistname  \
0  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   
1  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
2  9cc0cfd4d7d7885102480dd99e7a90d6                      Tiffany Page   
3  9cc0cfd4d7d7885102480dd99e7a90d6  Elvis Costello & The Attractions   
4  9cc0cfd4d7d7885102480dd99e7a90d6                    Elvis Costello   

                                           trackname    playlistname  
0               (The Angels Wanna Wear My) Red Shoes  HARD ROCK 2010  
1  (What's So Funny 'Bout) Peace, Love And Unders...  HARD ROCK 2010  
2                                   7 Years Too Late  HARD ROCK 2010  
3                              Accidents Will Happen  HARD ROCK 2010  
4                                             Alison  HARD ROCK 2010

In [13]:
# Try to check the artistname column
data["artistname"]

0                            Elvis Costello
1          Elvis Costello & The Attractions
2                              Tiffany Page
3          Elvis Costello & The Attractions
4                            Elvis Costello
                         ...               
1048570                            Dr. John
1048571                      Punch Brothers
1048572                      Chicago Farmer
1048573                 Henrik Freischlader
1048574                      Curtis Harding
Name: artistname, Length: 1047882, dtype: object

Now it's time to see if there is nulls in the data

In [14]:
# Get the info on the data
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1047882 entries, 0 to 1048574
Data columns (total 4 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   user_id       1047881 non-null  object
 1   artistname    1045679 non-null  object
 2   trackname     1047872 non-null  object
 3   playlistname  1047807 non-null  object
dtypes: object(4)
memory usage: 40.0+ MB


In [15]:
# Get the number of rows in the data
rows = data.shape[0]
# Get the number of nulls in the user_id column
nulls_user_id = rows - (rows - data["user_id"].isna().sum())
# Get the number of nulls in the artistname column
nulls_artist_name = rows - (rows - data["artistname"].isna().sum())
# Get the number of nulls in the trackname column
nulls_track_name = rows - (rows - data["trackname"].isna().sum())
# Get the number of nulls in the playlistname column
nulls_playlist_name = rows - (rows - data["playlistname"].isna().sum())

# Print off some useful info
print(f"There is {nulls_user_id} nulls in user_id \n")
print(f"There is {nulls_artist_name} nulls in artistname \n")
print(f"There is {nulls_track_name} nulls in trackname \n")
print(f"There is {nulls_playlist_name} nulls in playlistname \n")
print(f"There is {nulls_user_id+nulls_artist_name+nulls_track_name+nulls_playlist_name} all together")

There is 1 nulls in user_id 

There is 2203 nulls in artistname 

There is 10 nulls in trackname 

There is 75 nulls in playlistname 

There is 2289 all together


So, all together there is 2289 nulls in the data. And most of the nulls are in the artist name column. It would be very difficult to impute the data (mostly time consuming). For that reason I will be dropping the rows with nulls in them.

In [16]:
# Drop all the rows with nulls
data.dropna(inplace = True)

In [17]:
# Get the info on the data
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1045599 entries, 0 to 1048574
Data columns (total 4 columns):
 #   Column        Non-Null Count    Dtype 
---  ------        --------------    ----- 
 0   user_id       1045599 non-null  object
 1   artistname    1045599 non-null  object
 2   trackname     1045599 non-null  object
 3   playlistname  1045599 non-null  object
dtypes: object(4)
memory usage: 39.9+ MB


Nice and clean data!

Now to save the data as a .csv file.

In [18]:
# This is the name of the data 
dataname = r"\cleandata.csv"

# This is the directory of where im going to save the data
datadir = r"C:\Users\Alexander\Documents\Data Science Boot Camp\Capstone\Copys\Clean data"

# Saves the data to the directory specified 
data.to_csv(datadir+dataname)